crypto/cl/raw_creds.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2017 XLAB d.o.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package cl
+
+import (
+	"fmt"
+	"math/big"
+)
+
+// Attribute is part of a credential (RawCredential). In the case of digital identity credential,
+// attributes could be for example Name, Gender, Date of Birth. In the case of a credential allowing
+// access to some internet service (like electronic newspaper), attributes could be
+// Type (for example only news related to politics) of the service and Date of Expiration.
+type Attribute struct {
+	Index int
+	Name  string
+	Type  string // currently only "string" and "int" types are supported
+	Known bool
+	Value *big.Int // this value is set by SetAttributeValue and is used internally in CL scheme
+}
+
+func NewAttribute(index int, name, attrType string, known bool, value *big.Int) *Attribute {
+	return &Attribute{
+		Index: index,
+		Name:  name,
+		Type:  attrType,
+		Known: known,
+		Value: value,
+	}
+}
+
+// RawCredential presents credential as to be used by application that executes CL scheme to prove possesion
+// of an anonymous credential.
+type RawCredential struct {
+	attributes      map[int]*Attribute
+	attrNameToIndex map[string]int
+}
+
+func NewRawCredential() *RawCredential {
+	return &RawCredential{
+		attributes:      make(map[int]*Attribute),
+		attrNameToIndex: make(map[string]int),
+	}
+}
+
+func (c *RawCredential) InsertAttribute(index int, name, attrType string, known bool) {
+	c.attrNameToIndex[name] = index
+	c.attributes[index] = NewAttribute(index, name, attrType, known, nil)
+}
+
+func (c *RawCredential) AddAttribute(name, attrType string, known bool, value string) error {
```

Comment:
I think we could further simplify `AddAttribute`.

If we support int and string attributes, I think it's a good idea to offer `AddStringAttribute` and `AddIntAttribute`. As an end-user app, I would love to be able to do something like this:

````go
_ := c.AddStringAttribute("Name", "myname") // for a known string attribute
_ := c.AddIntAttribute("Age", 30).Hide() // for a hidden int attribute
````
The above example assumes that all attributes are known by default, and we can control their visibility with a method like `Hide()`. This way, we're left with an even more intuitive API, since we only need to provide attribute name and value.